File: 프로그래머스/2/12946. 하노이의 탑/하노이의 탑.py

def solution(n):
    answer = []
    
    def hanoi(n,before,tmp,after):
        if n==1:
            answer.append([before,after])
        else:
            hanoi(n-1,before,after,tmp)
            answer.append([before,after])
            hanoi(n-1,tmp,before,after)
    
    hanoi(n,1,2,3)
    return answer

File: 프로그래머스/2/12946. 하노이의 탑/하노이의 탑.java
import java.util.*;

class Solution {
    private List<int[]> list = new ArrayList<>();
    public int[][] solution(int n) {
        list = new ArrayList<>();
        
        hanoi(n, 1, 2, 3);

        int[][] answer = new int[list.size()][2];
        for (int i = 0; i < list.size(); i++) {
            answer[i] = list.get(i);
        }

        return answer;
    }
    
    private void hanoi(int n, int start, int mid, int end) {
        if (n == 1) {
            list.add(new int[]{start, end});
            return;
        }

        hanoi(n - 1, start, end, mid);
        list.add(new int[]{start, end});
        hanoi(n - 1, mid, start, end);
    }
}